Add course bundle spec and examples

joncass opened the pull request:
Here we "finalize" the course bundle spec and give a couple examples. I based the spec off of the draft quest bundle spec / examples that existed in the repo, as well as some offline thinking and discussions.

Looking for feedback on anything/everything, with particular questions around:
- The name `course`
- How to fully qualify labs and quizzes

Leave comments here and we can try to resolve things through discussion on here. If need be we can set up a separate meeting. I'm hoping to have this merged by the end of the week.

Thanks!

joncass commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Course bundle spec](https://trello.com/c/aOBEZq1t/2621-course-bundle-spec)

joncass commented:
After further discussion, we are hoping to use this spec for the *future* combined ILT/SPL concept. We will create separate GameTemplate and LearningPath specs (which will likely for now be identical to this one but have separate names to support unbundling into separate models and future enhancements that only apply to one or the other)

supersam654 commented on course-bundle-spec.md:
```diff
@@ -0,0 +1,125 @@
+  - activity_options:
+    - type: resource
+      content: intro-video
+    prompt:
```
@jgan may want to chime in here, but prompts were supposed to be displayed when a student is trying to decide which branch in a quest they want to take (either pick one of a few labs or skip an optional lab).

joncass replied:
Yeah in the quest bundle stuff I was working off of, they were used as a prompt when picking between labs. I figured I would make them more generally available to any step, to either be used in the cases Sam lists, or any other minor interstitial / preamble.

joncass replied:
After further conversation, we have decided to leave prompts to be used as described above.

joncass commented on course-bundle-spec.md:
```diff
@@ -0,0 +1,125 @@
+
+# Course Attributes
+id: gcp-intro-course
+default_locale: en
```
Modality: my understanding is that we want courses in general to be usable both as Instructor-Led and Self-Paced. Do you think there should be an intended modality, but the other is always available? Or maybe two attributes: `instructor_led` and `self_paced` and people could set one or both of them?

Course Code: Can that just be the `id`?

Level: Oy, this again :-p. Yeah we could add one - would you expect it to be an enum, or freetext?

joncass replied:
We agreed:
- Modality will not be included, since we want the content to be used in either context
- People can make the `id` their course code if they want
- We will add level: 1, 2, 3, 4

joncass commented on course-bundle-spec.md:
```diff
@@ -0,0 +1,125 @@
+objectives:
+  locales:
+    en: |
+      <p>This course will teach people</p>
```
Yeah I was intending to use the [instruction-html-spec](https://github.com/CloudVLab/qwiklabs-content-bundle-spec/blob/master/instruction-html-spec.md) for that, but now that I'm looking at it I think that's too broad. So we can create a separate whitelist. Do you have opinions on what should be in that?

supersam654 replied:
I propose: `p a ul ol li em strong` and perhaps `code pre`. The only attributes would be things that make `a` tags work which would be `title href` I believe.

Note that we'll need to be careful in Git Whisperer if people are allowed to write these fields as Markdown (with the expectation that GW will convert them to HTML).

As a separate issue entirely, we need to decide which fields can have HTML to begin with. I think all of the example fields with HTML are fine. I'm wondering if any of the example fields without HTML should also be included.

joncass commented on course-bundle-spec.md:
```diff
@@ -0,0 +1,125 @@
+    optional: true
+
+  - activity_options:
+    - type: lab
```
I agree with Sam - that doesn't really feel like a content thing, but I don't know too much about hot labs...

joncass commented on course-bundle-spec.md:
```diff
@@ -0,0 +1,125 @@
+---------------- | -------- | ----------- | -----------------------------------------
+activity_options | ✓        | array       | `activity_options` is an array of dictionaries with the format:
+-- type          | ✓        | enum        | One of `lab`, `quiz`, `resource`
+-- content       | ✓        | string      | Reference to the unique identifier for the activity. TODO: How to fully qualify `id` on labs and quizzes?
```
It can be ironed out later, but if people have opinions on it, I would love to have those stated here

supersam654 replied:
I'd like to limit what characters are allowed in a lab id (perhaps `[a-z0-9-]`) and then allow "external" labs to be namespaced (maybe `namespace/lab-id`) to reference something outside of the repo or outside of the library.

Other than restricting the valid characters in an id, I'm fine punting on the specifics of namespacing until later. It will probably be fairly abstract until Alexandria gets closer to completion anyway.

joncass commented on examples/course-robust/qwiklabs.yaml:
```diff
@@ -0,0 +1,150 @@
+objectives:
+  locales:
+    en: |
+      <p>This course will teach people</p>
```
Agreed. See my comment to Steve above. Do you have an opinion on what we should allow?

joncass replied:
Conversation continues above. Marking this resolved.

joncass commented on examples/course-robust/qwiklabs.yaml:
```diff
@@ -0,0 +1,150 @@
+        es: "./resources/storage_options_es.pdf"
+
+steps:
+  - activity_options:
```
Heh, yeah, I had the same conversation with Steve. `steps` is actually an array, not a hash, so `activity_options` is not the key describing the `step` (that is inferred from the fact that the `step` is an element in an array called `steps`), but is rather a key in a `step` hash declaring the activities that can be picked between for the given `step`. Does that clarify? Do you think I could make that clearer in the docs?

joncass commented on examples/course-robust/qwiklabs.yaml:
```diff
@@ -0,0 +1,150 @@
+      content: intro-to-gcp
+
+  - activity_options:
+    - type: resource
```
I see your point, but I think that the `resource` needs to declare a `type` (to make the `resources` easier to read), and I'd rather not have that duplicated here. So I lean towards leaving it this way. Curious what other people think.

joncass commented on course-bundle-spec.md:
```diff
@@ -0,0 +1,125 @@
+
+### Steps
+
+The meat of a course is an ordered list of steps defining what a learner needs to do to complete the course. A step consists of a set of one or more activity options, along with some metadata. Valid activity types are:
```
I was thinking of this as what Coursera calls a `module` and then we can build what Coursera calls a `course` out of these things. We should discuss further.

joncass replied:
After further discussion, we will be adding modules. I will make a commit for this shortly.

joncass commented on course-bundle-spec.md:
```diff
@@ -0,0 +1,125 @@
+  locales:
+    en: <p>A basic understanding of serverless architecture is recommended but not required.</p>
+
+tags: [sample, life-changing, gcp]
```
How do you think we should have authors specify those? Should they use the IDs from [this spreadsheet](https://docs.google.com/spreadsheets/d/1hUUch85HBRsRJsgRo9VCg0Pn7ZXi21sl6JU7VOr9LP8), or the strings from that spreadsheet, or what? Also should I link to that spreadsheet, or is there more "official" documentation somewhere?

joncass replied:
Added `gcp_product_tags`, `role_tags`, `domain_tags`

supersam654 commented on examples/course-minimal/qwiklabs.yaml:
```diff
@@ -33,11 +33,14 @@ prerequisites:
     en: <p>A basic understanding of serverless architecture is recommended but not required.</p>
 
 tags: [sample, life-changing, gcp]
+gcp_product_tags: ['bigquery']
```
Can we make this more generic (maybe `platform_product_tags`) so people can write `['bigquery']` or `['ec2', 's3']` or `['cuda', 'jupyter']`?

supersam654 replied:
Also, do we have a whitelist for these tags or can they be anything?

joncass replied:
Updated to `product_tags`. We have a whitelist - it is in the spreadsheet linked above in the attribute table.

supersam654 commented on examples/course-minimal/qwiklabs.yaml:
```diff
@@ -42,13 +42,17 @@ badge: gcp-intro-course-badge.png
-      content: gcp-quiz
+modules:
+  - title:
+      en: Module 1
```
Please nest this under `locales` to match other localized bundle content.

joncass replied:
Good catch. Will update shortly.

jgan commented on course-bundle-spec.md:
```diff
@@ -0,0 +1,164 @@
+schema_version: 1
+
+# Course Attributes
+id: gcp-intro-course
```
Ugh...

I know I'm responsible for the `id` field in the lab bundle spec, but I've always been hesitant about keeping an ID field in the course bundle itself instead of having it be externally defined via Alexandria or other publishing system.

@joncass Would you mind chatting with me for 15 minutes in front of a white board today, to talk about the implications of this?

joncass replied:
We decided to remove `id` from the bundle, and use the folder name as the author-defined `slug`.

eniskonuk commented on examples/course-robust/qwiklabs.yaml:
```diff
@@ -0,0 +1,174 @@
+
+resources:
+  - type: video
+    id: intro-video
```
where is the actual video resource (file)?

joncass commented:
@eniskonuk thanks for the comments!

- I am anticipating that the quiz spec will be added separately (in the not-too-distant future) by @rickcnagy.
- I don't have full context on what's in `classroom_template` nor what it is all used for, but hot labs and fraud protection sound more like runtime concerns than authoring/content concerns. I don't know what instructor modules are.
- I'd like to hear more about what you're picturing for course code and version string. @jgan may also have opinions there.

Let's discuss this morning.

joncass commented:
I discussed with Enis, and we decided:
- Okay for quiz spec to be added separately, but it shouldn't be too long before we get it in.
- Hot labs and fraud protection require more thought (I'm still leaning towards this being a classroom [/offering context] concern, but will discuss further with people).
- `instructor_resources` will be added.
- Course code is unneeded, as we will use folder names as slugs.
- Version string will be added, and we will update proposed website models accordingly (<-- this is likely to be controversial, and we may want a meeting to discuss [cc @jgan]).

joncass commented:
@eniskonuk - `version_string` will be handled at the locale-level; sorry that was unclear. And this bundle spec can handle all of the main features of those that I am aware of. We can expand it if/when we find pieces that we missed.

joncass commented:
Alright, based on yesterday afternoon's conversations I added two more commits:
- Rename `Course` to `CourseTemplate`, since we are standardizing on `FooTemplate` being the content set, and then `Foo` being the offered instance.
- Added versions as a locale dictionary, specifying that they should be considered decorators and not source of truth for revision history.

joncass commented:
Thanks for all the great discussion, everyone - I appreciate everyone taking the time to engage in this discussion. Merging now!
